kroxylicious-kms-provider-fortanix-dsm/src/main/java/io/kroxylicious/kms/provider/fortanix/dsm/session/ApiKeySessionProvider.java

```diff
@@ -0,0 +1,261 @@
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private static final TypeReference<SessionAuthResponse> SESSION_AUTH_RESPONSE = new TypeReference<>() {
+    };
+    private static final Duration HTTP_REQUEST_TIMEOUT = Duration.ofSeconds(10);
+    private static final Duration HTTP_CONNECT_TIMEOUT = Duration.ofSeconds(10);
+    private static final double DEFAULT_CREDENTIALS_LIFETIME_FACTOR = 0.80;
+    public static final String SESSION_AUTH_ENDPOINT = "/sys/v1/session/auth";
+
+    private final Clock systemClock;
+    private final AtomicReference<CompletableFuture<Session>> current = new AtomicReference<>();
+
+    private final AtomicLong tokenRefreshErrorCount = new AtomicLong();
+    private final Config config;
+    private final HttpClient client;
+
+    private final ScheduledExecutorService executorService;
+    @SuppressWarnings({ "java:S2245", "java:S2119" }) // Random used for backoff jitter, it does not need to be securely random.
+    private final ExponentialBackoff backoff = new ExponentialBackoff(500, 2, 60000, new Random().nextDouble());
+    private final Double lifetimeFactor;
+
+    /**
+     * Creates a session provider that uses an Api Key to authenticate.
+     *
+     * @param config config.
+     */
+    public ApiKeySessionProvider(@NonNull Config config) {
+        this(config, Clock.systemUTC());
+    }
+
+    @VisibleForTesting
+    ApiKeySessionProvider(@NonNull Config config, @NonNull Clock systemClock) {
+        Objects.requireNonNull(config);
+        Objects.requireNonNull(systemClock);
+        this.config = config;
+        this.systemClock = systemClock;
+        this.lifetimeFactor = Optional.ofNullable(config.apiKeyConfig().sessionLifetimeFactor()).orElse(DEFAULT_CREDENTIALS_LIFETIME_FACTOR);
+        this.executorService = Executors.newSingleThreadScheduledExecutor(r -> {
+            var thread = new Thread(r, ApiKeySessionProviderConfig.class.getName() + "thread");
+            thread.setDaemon(true);
+            return thread;
+        });
+        this.client = createClient();
+    }
+
+    private HttpClient createClient() {
+        var builder = HttpClient.newBuilder();
+        return builder
+                .followRedirects(HttpClient.Redirect.NORMAL)
+                .connectTimeout(HTTP_CONNECT_TIMEOUT)
+                .build();
+    }
+
+    @NonNull
+    @Override
+    public CompletionStage<Session> getSession() {
+        var newCredFuture = new CompletableFuture<Session>();
+        var witness = current.compareAndExchange(null, newCredFuture);
+        if (witness == null) {
+            // there's no current credential, let's create one
+            executorService.execute(() -> refreshCredential(newCredFuture));
+            return newCredFuture.minimalCompletionStage();
+        }
+        else if (isExpired(witness) || witness.isCompletedExceptionally()) {
```

Comment:
Swap the order, so I'm not left hanging on whether or not `isExpired` checks for exceptional completion.